File: app/src/main/java/com/giphy/browser/detail/GifDetailActivity.java
package com.giphy.browser.detail;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Animatable;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;

import com.facebook.drawee.backends.pipeline.Fresco;
import com.facebook.drawee.controller.BaseControllerListener;
import com.facebook.drawee.controller.ControllerListener;
import com.facebook.imagepipeline.image.ImageInfo;
import com.giphy.browser.R;
import com.giphy.browser.common.BaseActivity;
import com.giphy.browser.databinding.ActivityGifDetailBinding;

import java.util.Objects;

public class GifDetailActivity extends BaseActivity {

    private ActivityGifDetailBinding binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_gif_detail);

        final String webpUri = Objects.requireNonNull(getIntent().getStringExtra(WEBP_URI));

        setTitle(R.string.details);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        final ControllerListener listener = new BaseControllerListener<ImageInfo>() {
            @Override
            public void onIntermediateImageSet(String id, ImageInfo imageInfo) {
                updateViewSize(imageInfo);
            }

            @Override
            public void onFinalImageSet(String id, ImageInfo imageInfo, Animatable animatable) {
                updateViewSize(imageInfo);
            }

            private void updateViewSize(@Nullable ImageInfo imageInfo) {
                if (imageInfo != null) {
                    binding.gif.setAspectRatio((float) imageInfo.getWidth() / imageInfo.getHeight());
                }
            }
        };

        binding.gif.setController(
                Fresco.newDraweeControllerBuilder()
                        .setAutoPlayAnimations(true)
                        .setControllerListener(listener)
                        .setOldController(binding.gif.getController())
                        .setUri(webpUri)
                        .build());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private static final String WEBP_URI = "webp_uri";

    public static Intent newIntent(@NonNull Context context, @NonNull String webpUri) {
        return new Intent(context, GifDetailActivity.class)
                .putExtra(WEBP_URI, webpUri);
    }
}

File: app/src/main/java/com/giphy/browser/gif_detail/GifDetailActivity.kt
package com.giphy.browser.gif_detail

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.view.MenuItem
import androidx.annotation.ColorRes
import androidx.databinding.DataBindingUtil
import com.facebook.drawee.backends.pipeline.Fresco
import com.giphy.browser.R
import com.giphy.browser.common.BaseActivity
import com.giphy.browser.databinding.ActivityGifDetailBinding
import kotlinx.android.parcel.Parcelize
import java.util.*

class GifDetailActivity : BaseActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding: ActivityGifDetailBinding =
            DataBindingUtil.setContentView(this, R.layout.activity_gif_detail)

        val args = intent.getParcelableExtra(ARGS) as Args

        setTitle(R.string.details)
        Objects.requireNonNull(supportActionBar)!!.setDisplayHomeAsUpEnabled(true)

        binding.gif.aspectRatio = args.aspectRatio
        binding.gif.hierarchy.setPlaceholderImage(args.backgroundColor)
        binding.gif.controller = Fresco.newDraweeControllerBuilder()
            .setAutoPlayAnimations(true)
            .setOldController(binding.gif.controller)
            .setUri(args.webp)
            .build()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    @Parcelize
    data class Args(
        val webp: String,
        val aspectRatio: Float,
        @ColorRes val backgroundColor: Int
    ) : Parcelable

    companion object {
        private const val ARGS = "args"
        fun newIntent(context: Context, args: Args): Intent {
            return Intent(context, GifDetailActivity::class.java)
                .putExtra(ARGS, args)
        }
    }
}
